// @refinedev/supabase/dist/index.mjs
// src/index.ts
import { createClient } from "@supabase/supabase-js";

// src/types/index.ts
import { REALTIME_POSTGRES_CHANGES_LISTEN_EVENT } from "@supabase/supabase-js";
var liveTypes = {
  INSERT: "created",
  UPDATE: "updated",
  DELETE: "deleted",
  "*": "*"
};
var supabaseTypes = {
  created: REALTIME_POSTGRES_CHANGES_LISTEN_EVENT.INSERT,
  updated: REALTIME_POSTGRES_CHANGES_LISTEN_EVENT.UPDATE,
  deleted: REALTIME_POSTGRES_CHANGES_LISTEN_EVENT.DELETE,
  "*": REALTIME_POSTGRES_CHANGES_LISTEN_EVENT.ALL
};

// src/utils/mapOperator.ts
var mapOperator = (operator) => {
  switch (operator) {
    case "ne":
      return "neq";
    case "nin":
      return "not.in";
    case "contains":
      return "ilike";
    case "ncontains":
      return "not.ilike";
    case "containss":
      return "like";
    case "ncontainss":
      return "not.like";
    case "null":
      return "is";
    case "nnull":
      return "not.is";
    case "ina":
      return "cs";
    case "nina":
      return "not.cs";
    case "between":
    case "nbetween":
      throw Error(`Operator ${operator} is not supported`);
    default:
      return operator;
  }
};

// src/utils/generateFilter.ts
var generateFilter = (filter, query) => {
  switch (filter.operator) {
    case "eq":
      return query.eq(filter.field, filter.value);
    case "ne":
      return query.neq(filter.field, filter.value);
    case "in":
      return query.in(filter.field, filter.value);
    case "ina":
      return query.contains(filter.field, filter.value);
    case "nina":
      return query.not(
        filter.field,
        "cs",
        `{${filter.value.map((val) => `"${val}"`).join(",")}}`
      );
    case "gt":
      return query.gt(filter.field, filter.value);
    case "gte":
      return query.gte(filter.field, filter.value);
    case "lt":
      return query.lt(filter.field, filter.value);
    case "lte":
      return query.lte(filter.field, filter.value);
    case "between":
      if (filter.value.length !== 2) {
        throw new Error(
          `[@refinedev/supabase]: Unexpected length ${filter.value.length}. Between operator expects a range between 2 values.`
        );
      }
      return query.gte(filter.field, filter.value[0]).lte(filter.field, filter.value[1]);
    case "contains":
      return query.ilike(filter.field, `%${filter.value}%`);
    case "containss":
      return query.like(filter.field, `%${filter.value}%`);
    case "null":
      return query.is(filter.field, null);
    case "startswith":
      return query.ilike(filter.field, `${filter.value}%`);
    case "endswith":
      return query.ilike(filter.field, `%${filter.value}`);
    case "or": {
      const orSyntax = filter.value.map((item) => {
        if (item.operator !== "or" && item.operator !== "and" && "field" in item) {
          let value = item.value;
          if (item.operator === "ina" || item.operator === "nina") {
            value = `{${item.value.map((val) => `"${val}"`).join(",")}}`;
          }
          if (item.operator === "contains" || item.operator === "containss") {
            value = `"%${value}%"`;
          }
          if (item.operator === "startswith") {
            value = `${value}%`;
          }
          if (item.operator === "endswith") {
            value = `%${value}`;
          }
          if (item.operator === "in") {
            value = `(${item.value.map((val) => `"${val}"`).join(",")})`;
          }
          return `${item.field}.${mapOperator(item.operator)}.${value}`;
        }
        return;
      }).join(",");
      return query.or(orSyntax);
    }
    case "and":
      throw Error("Operator 'and' is not supported");
    default:
      return query.filter(
        filter.field,
        mapOperator(filter.operator),
        filter.value
      );
  }
};

// src/utils/handleError.ts
var handleError = (error) => {
  const customError = {
    ...error,
    message: error.message,
    statusCode: Number.parseInt(error.code)
  };
  return Promise.reject(customError);
};

// src/dataProvider/index.ts
var dataProvider = (supabaseClient) => {
  return {
    getList: async ({ resource, pagination, filters, sorters, meta }) => {
      const {
        currentPage = 1,
        pageSize = 10,
        mode = "server"
      } = pagination ?? {};
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).select((meta == null ? void 0 : meta.select) ?? "*", {
        count: (meta == null ? void 0 : meta.count) ?? "exact"
      });
      if (mode === "server") {
        query.range((currentPage - 1) * pageSize, currentPage * pageSize - 1);
      }
      sorters == null ? void 0 : sorters.map((item) => {
        const [foreignTable, field] = item.field.split(/\.(?=[^.]+$)/);
        if (foreignTable && field) {
          query.select((meta == null ? void 0 : meta.select) ?? `*, ${foreignTable}(${field})`).order(field, {
            ascending: item.order === "asc",
            foreignTable
          });
        } else {
          query.order(item.field, {
            ascending: item.order === "asc"
          });
        }
      });
      filters == null ? void 0 : filters.map((item) => {
        generateFilter(item, query);
      });
      const { data, count, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: data || [],
        total: count || 0
      };
    },
    getMany: async ({ resource, ids, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).select((meta == null ? void 0 : meta.select) ?? "*");
      if (meta == null ? void 0 : meta.idColumnName) {
        query.in(meta.idColumnName, ids);
      } else {
        query.in("id", ids);
      }
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: data || []
      };
    },
    create: async ({ resource, variables, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).insert(variables);
      query.select((meta == null ? void 0 : meta.select) ?? "*");
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: (data || [])[0]
      };
    },
    createMany: async ({ resource, variables, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).insert(variables);
      query.select((meta == null ? void 0 : meta.select) ?? "*");
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data
      };
    },
    update: async ({ resource, id, variables, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).update(variables);
      if (meta == null ? void 0 : meta.idColumnName) {
        query.eq(meta.idColumnName, id);
      } else {
        query.match({ id });
      }
      query.select((meta == null ? void 0 : meta.select) ?? "*");
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: (data || [])[0]
      };
    },
    updateMany: async ({ resource, ids, variables, meta }) => {
      const response = await Promise.all(
        ids.map(async (id) => {
          const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
          const query = client.from(resource).update(variables);
          if (meta == null ? void 0 : meta.idColumnName) {
            query.eq(meta.idColumnName, id);
          } else {
            query.match({ id });
          }
          query.select((meta == null ? void 0 : meta.select) ?? "*");
          const { data, error } = await query;
          if (error) {
            return handleError(error);
          }
          return (data || [])[0];
        })
      );
      return {
        data: response
      };
    },
    getOne: async ({ resource, id, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).select((meta == null ? void 0 : meta.select) ?? "*");
      if (meta == null ? void 0 : meta.idColumnName) {
        query.eq(meta.idColumnName, id);
      } else {
        query.match({ id });
      }
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: (data || [])[0]
      };
    },
    deleteOne: async ({ resource, id, meta }) => {
      const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
      const query = client.from(resource).delete();
      if (meta == null ? void 0 : meta.idColumnName) {
        query.eq(meta.idColumnName, id);
      } else {
        query.match({ id });
      }
      const { data, error } = await query;
      if (error) {
        return handleError(error);
      }
      return {
        data: (data || [])[0]
      };
    },
    deleteMany: async ({ resource, ids, meta }) => {
      const response = await Promise.all(
        ids.map(async (id) => {
          const client = (meta == null ? void 0 : meta.schema) ? supabaseClient.schema(meta.schema) : supabaseClient;
          const query = client.from(resource).delete();
          if (meta == null ? void 0 : meta.idColumnName) {
            query.eq(meta.idColumnName, id);
          } else {
            query.match({ id });
          }
          const { data, error } = await query;
          if (error) {
            return handleError(error);
          }
          return (data || [])[0];
        })
      );
      return {
        data: response
      };
    },
    getApiUrl: () => {
      throw Error("Not implemented on refine-supabase data provider.");
    },
    custom: () => {
      throw Error("Not implemented on refine-supabase data provider.");
    }
  };
};

// src/liveProvider/index.ts
var liveProvider = (supabaseClient) => {
  return {
    subscribe: ({
      channel,
      types,
      params,
      callback,
      meta
    }) => {
      var _a;
      const resource = channel.replace("resources/", "");
      const listener = (payload) => {
        var _a2;
        if (types.includes("*") || types.includes(liveTypes[payload.eventType])) {
          if (liveTypes[payload.eventType] !== "created" && (params == null ? void 0 : params.ids) !== void 0 && ((_a2 = payload.new) == null ? void 0 : _a2.id) !== void 0) {
            if (params.ids.map(String).includes(payload.new.id.toString())) {
              callback({
                channel,
                type: liveTypes[payload.eventType],
                date: new Date(payload.commit_timestamp),
                payload: payload.new
              });
            }
          } else {
            callback({
              channel,
              type: liveTypes[payload.eventType],
              date: new Date(payload.commit_timestamp),
              payload: payload.new
            });
          }
        }
      };
      const mapFilter = (filters) => {
        if (!filters || filters.length === 0) {
          return;
        }
        const mapped = filters.map((filter2) => {
          if ("field" in filter2) {
            return `${filter2.field}=${mapOperator(filter2.operator)}.${filter2.value}`;
          }
          return;
        }).filter((x) => Boolean(x));
        if (mapped.length === 0)
          return;
        if (mapped.length > 1) {
          console.warn(
            `[refine/supabase] Multiple filters are not supported for Supabase Realtime subscriptions. Using only the first filter: "${mapped[0]}".`
          );
        }
        return mapped[0];
      };
      const events = types.map((x) => supabaseTypes[x]).sort((a, b) => a.localeCompare(b));
      const filter = mapFilter(params == null ? void 0 : params.filters);
      const ch = `${channel}:${events.join("|")}${filter ? `:${filter}` : ""}`;
      let client = supabaseClient.channel(ch);
      for (let i = 0; i < events.length; i++) {
        client = client.on(
          "postgres_changes",
          {
            event: events[i],
            schema: (meta == null ? void 0 : meta.schema) || // @ts-expect-error TS2445 Property rest is protected and only accessible within class SupabaseClient and its subclasses.
            ((_a = supabaseClient == null ? void 0 : supabaseClient.rest) == null ? void 0 : _a.schemaName) || "public",
            table: resource,
            filter
          },
          listener
        );
      }
      return client.subscribe();
    },
    unsubscribe: async (channel) => {
      supabaseClient.removeChannel(channel);
    }
  };
};
export {
  createClient,
  dataProvider,
  generateFilter,
  handleError,
  liveProvider,
  liveTypes,
  mapOperator,
  supabaseTypes
};
